src/Aspire.Hosting/ApplicationModel/ResourceLoggerService.cs

```diff
@@ -244,9 +249,28 @@ void Log(LogLine log)
 
             try
             {
-                await foreach (var entry in channel.GetBatchesAsync(cancellationToken: cancellationToken).ConfigureAwait(false))
+                var lineCount = 0;
+                while (await channel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
                 {
-                    yield return entry;
+                    var batch = new List<LogLine>();
+                    while (channel.Reader.TryRead(out var log))
+                    {
+                        if (lineCount > backlogSnapshot.Length)
+                        {
+                            // We're beyond the number of lines in the backlog snapshot so no need to check for dupes anymore
+                            batch.Add(log);
+                        }
+                        else
+                        {
+                            // We need to check to ensure we haven't already returned this line via the backlog snapshot we took
+                            if (!backlogSnapshot.Contains(log))
```

Comment:
Do IIUC that even if the same exact message string would be logged the `LogLine` instance here would be different because it would always have a monotonically increasing `lineNumber`? And so it wouldn't skip repeated message strings when they are in the snapshot, and on the reader?

Would it be useful to add a test for repeated strings being logged?

I don't know much here so my question might be non-sensical!

Reply:
> Do IIUC that even if the same exact message string would be logged the LogLine instance here would be different because it would always have a monotonically increasing lineNumber? And so it wouldn't skip repeated message strings when they are in the snapshot, and on the reader?

Correct. Because we are using `record` for LogLine, it automatically gets an `Equals` override that compares all of its properties for equality. So if the line numbers are different, they won't equal.

Reply:
If we thought we needed a faster check here, instead of `Contains`, I guess we could just check for `LineNumber`, or even check if the `LineNumber` is greater than the `backlogSnapshot.Length`. But that might be premature optimization. The current code seems sufficient.